// trip-service/dist/dto/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BookingStatus = exports.BookingType = exports.TripStatus = exports.UpdateBookingDto = exports.CreateBookingDto = exports.UpdateTripDto = exports.CreateTripDto = void 0;
var trip_dto_1 = require("./trip.dto");
Object.defineProperty(exports, "CreateTripDto", { enumerable: true, get: function () { return trip_dto_1.CreateTripDto; } });
Object.defineProperty(exports, "UpdateTripDto", { enumerable: true, get: function () { return trip_dto_1.UpdateTripDto; } });
var booking_dto_1 = require("./booking.dto");
Object.defineProperty(exports, "CreateBookingDto", { enumerable: true, get: function () { return booking_dto_1.CreateBookingDto; } });
Object.defineProperty(exports, "UpdateBookingDto", { enumerable: true, get: function () { return booking_dto_1.UpdateBookingDto; } });
var types_1 = require("../types");
Object.defineProperty(exports, "TripStatus", { enumerable: true, get: function () { return types_1.TripStatus; } });
Object.defineProperty(exports, "BookingType", { enumerable: true, get: function () { return types_1.BookingType; } });
Object.defineProperty(exports, "BookingStatus", { enumerable: true, get: function () { return types_1.BookingStatus; } });
